cmd/microcloud/session.go

```diff
@@ -0,0 +1,166 @@
+	if err != nil {
+		return fmt.Errorf("Failed to send session start: %w", err)
+	}
+
+	err = gw.ReceiveCombined(&session)
+	if err != nil {
+		return fmt.Errorf("Failed to read session reply: %w", err)
+	}
+
+	if !c.autoSetup {
+		fmt.Printf("Enter the passphrase on systems that you want to join the cluster:\n\n %s\n\n", session.Passphrase)
+		fmt.Println("Waiting for systems to send their intent to join ...")
+	}
+
+	confirmedIntents, err := c.askJoinIntents(gw, expectedSystems)
+	if err != nil {
+		return err
+	}
+
+	if !c.autoSetup {
+		cloud := sh.Services[types.MicroCloud].(*service.CloudService)
+		cert, err := cloud.Certificate()
+		if err != nil {
+			return err
+		}
+
+		fingerprint, err := c.shortFingerprint(cert.Fingerprint())
+		if err != nil {
+			return fmt.Errorf("Failed to shorten fingerprint: %w", err)
+		}
+
+		fmt.Printf(" Using fingerprint %q to confirm the join intents\n", fingerprint)
+	}
+
+	err = gw.Write(types.Session{
+		ConfirmedIntents: confirmedIntents,
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to send join intents: %w", err)
+	}
+
+	err = gw.ReceiveCombined(&session)
+	if err != nil {
+		return fmt.Errorf("Failed to read confirmation errors: %w", err)
+	}
+
+	if !session.Accepted {
+		return fmt.Errorf("Join confirmations didn't get accepted on all systems")
+	}
+
+	for _, joinIntent := range confirmedIntents {
+		certBlock, _ := pem.Decode([]byte(joinIntent.Certificate))
+		if certBlock == nil {
+			return fmt.Errorf("Invalid certificate file")
+		}
+
+		remoteCert, err := x509.ParseCertificate(certBlock.Bytes)
+		if err != nil {
+			return fmt.Errorf("Failed to parse certificate: %w", err)
+		}
+
+		// Register init system
+		c.systems[joinIntent.Name] = InitSystem{
+			ServerInfo: mdns.ServerInfo{
+				Version:  joinIntent.Version,
+				Name:     joinIntent.Name,
+				Address:  joinIntent.Address,
+				Services: joinIntent.Services,
+				// Store the peers certificate to allow mTLS server validation
+				// for requests after the trust establishment.
+				Certificate: remoteCert,
+			},
+		}
+	}
+
+	// Add a space between the CLI and the response.
+	fmt.Println("")
+
+	if !c.autoSetup {
+		for _, info := range c.systems {
+			fmt.Printf(" Selected %q at %q\n", info.ServerInfo.Name, info.ServerInfo.Address)
+		}
+	}
+
+	// Add a space between the CLI and the response.
+	fmt.Println("")
+
+	return nil
+}
+
+func (c *initConfig) joiningSession(gw *cloudClient.WebsocketGateway, sh *service.Handler, services []types.ServiceType, target string, passphrase string) error {
+	session := types.Session{
+		Passphrase:    passphrase,
+		Address:       sh.Address,
+		TargetAddress: target,
+		Interface:     c.lookupIface.Name,
+		Services:      services,
+		LookupTimeout: c.lookupTimeout,
+	}
+
+	err := gw.Write(session)
+	if err != nil {
+		return fmt.Errorf("Failed to send session start: %w", err)
+	}
+
+	if !c.autoSetup && target == "" {
+		fmt.Println("Searching an eligible system ...")
+	}
```

Comment:
Is this line necessary?

Reply:
In case you rely on mDNS and haven't specified `--target` we have to wait for any mDNS server to respond to our query. If this is taking a bit longer or if there is no `microcloud init` running on another node you will get something like this:

```
microcloud join --lookup-timeout 5
...
Searching an eligible system ...
Error: Failed to find an eligible system: Failed lookup: context deadline exceeded
```

Or the session timeout kicks in whatever happens first.

Reply:
Should be "Searching _for_ an eligible system ..." right?

Also maybe `s/system/initiator/`?

Reply:
Good point, updated the `Searching ` line.

In case of interactive mode should we introduce terms like initiator/joiner? I think it makes it easier to follow if we just assume those things under the hood.

Whereas in case of preseed I think it's totally fine to introduce these terms as you need to understand when to set either `initiator` or `initiator_address` in the preseed file.